lib/mongo/collection.ex

```diff
@@ -545,133 +545,92 @@ defmodule Mongo.Collection do
-            end
-
-          false ->
-            def new() do
-              %__MODULE__{unquote_splicing(Collection.struct_args(args))}
-            end
+        def new do
+          if @timestamps != [],
+            do: new_timestamps(%__MODULE__{unquote_splicing(Collection.struct_args(args))}),
+            else: %__MODULE__{unquote_splicing(Collection.struct_args(args))}
         end
       end
 
     load_function =
       quote unquote: false do
-        attribute_names = @attributes |> Enum.map(&elem(&1, 0))
+        attribute_names = Enum.map(@attributes, fn {name, opts} -> {name, opts[:name]} end)
 
         embed_ones =
           @embed_ones
           |> Enum.filter(fn {_name, mod, _opts} -> Collection.has_load_function?(mod) end)
-          |> Enum.map(fn {name, mod, _opts} -> {name, mod} end)
+          |> Enum.map(fn {name, mod, opts} -> {name, {opts[:name], mod}} end)
 
         embed_manys =
           @embed_manys
           |> Enum.filter(fn {_name, mod, _opts} -> Collection.has_load_function?(mod) end)
-          |> Enum.map(fn {name, mod, _opts} -> {name, mod} end)
+          |> Enum.map(fn {name, mod, opts} -> {name, {opts[:name], mod}} end)
 
         def load(map, use_atoms \\ false)
+        def load(nil, _use_atoms), do: nil
+        def load(xs, use_atoms) when is_list(xs), do: Enum.map(xs, fn map -> load(map, use_atoms) end)
 
-        def load(nil, _use_atoms) do
-          nil
-        end
-
-        def load(xs, use_atoms) when is_list(xs) do
-          Enum.map(xs, fn map -> load(map, use_atoms) end)
-        end
-
-        def load(map, false) when is_map(map) do
-          struct =
-            Enum.reduce(
-              unquote(attribute_names),
-              %__MODULE__{},
-              fn name, result ->
-                Map.put(result, name, map[Atom.to_string(name)])
-              end
-            )
-
-          struct =
-            unquote(embed_ones)
-            |> Enum.map(fn {name, mod} -> {name, mod.load(map[Atom.to_string(name)])} end)
-            |> Enum.reduce(struct, fn {name, doc}, acc -> Map.put(acc, name, doc) end)
-
-          unquote(embed_manys)
-          |> Enum.map(fn {name, mod} -> {name, mod.load(map[Atom.to_string(name)])} end)
-          |> Enum.reduce(struct, fn {name, doc}, acc -> Map.put(acc, name, doc) end)
-          |> @after_load_fun.()
-        end
-
-        def load(map, true) when is_map(map) do
-          struct =
-            Enum.reduce(
-              unquote(attribute_names),
-              %__MODULE__{},
-              fn name, result ->
-                Map.put(result, name, map[name])
-              end
-            )
+        def load(map, use_atoms) when is_map(map) do
+          map = if use_atoms, do: map, else: Enum.into(map, %{}, fn {name, doc} -> {String.to_atom(name), doc} end)
+          struct = Enum.reduce(unquote(attribute_names), %__MODULE__{}, fn {name, src_name}, result -> Map.put(result, name, map[src_name]) end)
```

Comment:
Does this work? I think the if-statement stuff is wrong. And if I rerun the benchee, the penalty still exists:
```
Name             ips        average  deviation         median         99th %
load-2        1.53 M      653.80 ns  ±2775.75%         583 ns         750 ns
load          1.21 M      825.85 ns  ±2234.43%         708 ns         917 ns

Comparison:
load-2        1.53 M
load          1.21 M - 1.26x slower +172.05 ns
```

The `use_atom` is a flag to select the type of the keys in the *source doc*. The keys of the documents, that are returned from the DB, have the types of binary. The line converts the keys of the destination *struct* into atoms, while they are already atoms.

Reply:
Ah...I understand it now: if use_atoms, you convert all keys from the source map to atoms. That is dangerous because a source map contains a lot of different keys and you would create a memory leak.

Reply:
I changed it to precompiled string keys

Reply:
https://github.com/zookzook/elixir-mongodb-driver/pull/149/commits/b43d17187104ea04b798bed8dc7da47f454bcbde